### PR TITLE
Add a DiffStat type, methods, and tests

### PR DESCRIPTION
**Pull request by kshyatt**
Create a DiffStat type so that we can get more information
about GitDiffs. Add a show method for this type and GitDiff.
Also add methods to get the number of files changed/insertions/
deletions in the diff.

### CODE REVIEW

File: base/libgit2/diff.jl [hunk 1]
```diff
@@ -34,6 +34,26 @@ function diff_tree(repo::GitRepo, oldtree::GitTree, newtree::GitTree)
     return GitDiff(repo, diff_ptr_ptr[])
 end
 
+function GitDiffStats(diff::GitDiff)
```
Comment (ararslan) [inline base/libgit2/diff.jl]: Does a type already exist for this?

Comment (kshyatt) [inline base/libgit2/diff.jl]: I don't think so. I grepped for it. This is the way you're supposed to extract stats from the diff, I think, according to the API docs.

Comment (ararslan) [inline base/libgit2/diff.jl]: If there isn't a `GitDiffStats` type defined then you can't use it as an annotation in the method definitions below though

Comment (ararslan) [inline base/libgit2/diff.jl]: HAH disregard. The type is defined in types.jl as part of this PR. _facepalm_

**Review by ararslan — APPROVED** [review]
Nice, I like this a lot

**Comment by kshyatt** [comment]
Can someone weigh in on whether this new type will be finalized correctly? @simonbyrne or @tkelman or @yuyichao I don't want to introduce a bug and I'm not super-familiar with how this works. There is a method [git_diff_stats_free](https://libgit2.github.com/libgit2/#HEAD/group/diff/git_diff_stats_free) which should be called from https://github.com/JuliaLang/julia/blob/master/base/libgit2/types.jl#L539 properly, right?

**Review by StefanKarpinski — APPROVED** [review]
Here's my analysis of the finalizer situation:

1. Look at the diff with whitespace ignored: https://github.com/JuliaLang/julia/pull/20730/files?w=1
2. It adds the `GitDiffStats` type, via metaprogramming
3. The `GitDiffStats` type  is defined here: https://github.com/JuliaLang/julia/blob/0b2130e628da2ab23b4eb7cd5ff08e6e42c60bd0/base/libgit2/types.jl#L503-L516, including adding a finalizer that calls `Base.close` on the object.
4. That `close` method is defined here: https://github.com/JuliaLang/julia/blob/0b2130e628da2ab23b4eb7cd5ff08e6e42c60bd0/base/libgit2/types.jl#L549-L558. It calls the appropriate `git_*_free` function on the object if the pointer is non-null, which presumably frees its memory correctly. It then sets the pointer to `null` so that no one tries to free it again.
5. Finally, if the total refcount to `libgit2` objects is zero, it calls `git_libgit2_shutdown`.

This all looks kosher to me.

File: base/libgit2/types.jl [hunk 2]
```diff
@@ -471,23 +471,24 @@ Base.isempty(obj::AbstractGitObject) = (obj.ptr == C_NULL)
...
-    (:GitStatus,     :GitRepo,  :AbstractGitObject, :git_status_list),
-    (:GitBranchIter, :GitRepo,  :AbstractGitObject, :git_branch_iterator),
-    (:GitUnknownObject,  :GitRepo,  :GitObject,         :git_object),
-    (:GitCommit,     :GitRepo,  :GitObject,         :git_commit),
-    (:GitBlob,       :GitRepo,  :GitObject,         :git_blob),
-    (:GitTree,       :GitRepo,  :GitObject,         :git_tree),
-    (:GitTag,        :GitRepo,  :GitObject,         :git_tag)]
+    (:GitRepo,          nothing,   :AbstractGitObject, :git_repository),
+    (:GitTreeEntry,     nothing,   :AbstractGitObject, :git_tree_entry),
+    (:GitDiffStats,     nothing,   :AbstractGitObject, :git_diff_stats),
```
Comment (simonbyrne) [inline base/libgit2/types.jl]: In general, it should probably keep a reference to the `GitRepo` unless it is known to be safe not to.

**Pull request merged**